hax/hax/motr/__init__.py

```diff
@@ -355,19 +360,35 @@ def ha_obj_state(st):
                                                             kv_cache=kv_cache)
         if not notes:
             return []
+        message_ids = self._ha_broadcast(notes, broadcast_hax_only)
+
+        return message_ids
+
+    def _ha_broadcast(self, notes: List[HaNoteStruct],
+                      broadcast_hax_only: bool) -> List[MessageId]:
+        idx = 0
         message_ids: List[MessageId] = []
-        LOG.debug('broadcast ha states phase %s', broadcast_hax_only)
-        if broadcast_hax_only:
-            hax_endpoint = self.consul_util.get_hax_endpoint()
-            message_ids = self._ffi.ha_broadcast_hax_only(
-                self._ha_ctx, make_array(HaNoteStruct, notes), len(notes),
-                make_c_str(hax_endpoint))
-        else:
-            message_ids = self._ffi.ha_broadcast(
-                self._ha_ctx, make_array(HaNoteStruct, notes), len(notes))
-        LOG.debug(
-            'Broadcast HA state complete with the following message_ids = %s',
-            message_ids)
+        total_len = len(notes)
+        LOG.debug('Broadcast nodes total len %d', total_len)
+        while notes:
+            notes_to_send = notes[0:MAX_MOTR_NVEC_UPDATE_SZ]
+            notes_to_send_len = len(notes_to_send)
+            if broadcast_hax_only:
+                hax_endpoint = self.consul_util.get_hax_endpoint()
+                message_ids = self._ffi.ha_broadcast_hax_only(
+                    self._ha_ctx, make_array(HaNoteStruct, notes_to_send),
+                    notes_to_send_len,
+                    make_c_str(hax_endpoint))
+            else:
+                message_ids = self._ffi.ha_broadcast(
+                    self._ha_ctx, make_array(HaNoteStruct, notes_to_send),
+                    notes_to_send_len)
+            LOG.debug(
+                'Broadcast HA state complete, message_ids = %s',
+                message_ids)
+            idx += notes_to_send_len
+            notes = notes[idx:total_len]
```

Comment:
I'm afraid this is the same mistake that I saw in the first version of the algorithm. I guess that `idx` and `notes` will become out of sync after the second iteration already.

## Example
For simplicity, let's consider `MAX_MOTR_NVEC_UPDATE_SZ = 2` and `notes = [0, 1, 2, 3, 4, 5]`
### Iteration 1
```python
total_len = 6
notes_to_send = [0, 1]
notes_to_send_len = len([0, 1]) = 2
idx = 2
notes = notes[2: 6] = [2, 3, 4, 5]
```

### Iteration 2
```python
notes_to_send = [2, 3]
notes_to_send_len = len([2, 3]) = 2
idx = 2 + 2 = 4
notes = notes[4: 6] = []
```
As a result, messages [4, 5, 6] will not be sent.

# Solution

```suggestion
            notes = notes[MAX_MOTR_NVEC_UPDATE_SZ:]
```
And `idx` becomes unneeded.

Reply:
you are right, don't need idx, my bad.

Reply:
Updated patch... testing in-progress...

Reply:
testing done, updated with results.